### PR TITLE
Add endpoints to modify annotations on existing Clusters and Installations

### PR DESCRIPTION
**Pull request by Szymongib**
#### Summary

Add endpoints to add and remove Annotations from existing Clusters and Installations.
Endpoints have the following structure:
- To add set of Annotations to the Cluster:
   `POST /api/cluster/[CLUSTER_ID]/annotations` 
   ```json
   {
       "annotations": [
           "test",  "multi-tenant"
       ]
   }
   ```
- To delete an Annotation from the Cluster
   `DELETE /api/cluster/[CLUSTER_ID]/annotation/[ANNOTATION_NAME]`
- To add set of Annotations to Installation
  `POST /api/installation/[INSTALLATION_ID]/annotations` 
   ```json
   {
       "annotations": [
           "test",  "multi-tenant"
       ]
   }
   ```
- To delete an Annotation from the Installation
   `DELETE /api/installation/[INSTALLATION_ID]/annotation/[ANNOTATION_NAME]`

To preserve Annotations state matching the scheduling state, the following constraints apply to the operations:
- When deleting Cluster Annotation - the Annotation which is being deleted cannot be present on any Installation currently scheduled on that Cluster.
- When adding Annotation to the Installation - the Annotation needs to be present in every Cluster on which the Installation is scheduled.

Commands added to CLI:
- To add Annotations to the Cluster 
   ```
   cloud cluster annotation add --cluster=[CLUSTER_ID] --annotation [ANNOTATION_1] --annotation [ANNOTATION_2]
   ```
- To delete Annotation from the Cluster 
   ```
   cloud cluster annotation delete --cluster=[CLUSTER_ID] --annotation [ANNOTATION_NAME]
   ```
- To add Annotations to the Installation 
   ```
   cloud installation annotation add --installation=[CLUSTER_ID] --annotation [ANNOTATION_1] --annotation [ANNOTATION_2]
   ```
- To delete Annotation from the Installation 
   ```
   cloud installation annotation delete --installation=[CLUSTER_ID] --annotation [ANNOTATION_NAME]
   ```

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-29474

#### Release Note

```release-note
Add endpoints to modify annotations on existing Clusters and Installations
```

### CODE REVIEW

**Comment by jwilander** [comment]
This ready for reviewers @Szymongib ?

**Comment by Szymongib** [comment]
@jwilander I need to fix conflicts and update the docs, after merging the previous PR, then it is ready.

**Review by jwilander — APPROVED** [review]
Looks good 👍

File: cmd/cloud/cluster_annotation.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	"github.com/spf13/cobra"
+)
+
+func init() {
+	clusterAnnotationAddCmd.Flags().StringArray("annotation", []string{}, "Additional annotations for the cluster. Accepts multiple values, for example: '... --annotation abc --annotation def'")
+	clusterAnnotationAddCmd.Flags().String("cluster", "", "The id of the cluster to be annotated.")
+	clusterAnnotationAddCmd.MarkFlagRequired("cluster")
+	clusterAnnotationAddCmd.MarkFlagRequired("annotation")
+
+	clusterAnnotationDeleteCmd.Flags().String("annotation", "", "Name of the annotation to be removed from the Cluster.")
```
Comment (gabrieljackson) [inline cmd/cloud/cluster_annotation.go:19]:
```suggestion
	clusterAnnotationDeleteCmd.Flags().String("annotation", "", "Name of the annotation to be removed from the cluster.")
```

File: cmd/cloud/installation_annotation.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+var installationAnnotationCmd = &cobra.Command{
+	Use:   "annotation",
+	Short: "Manipulate annotations of installations managed by the provisioning server.",
+}
+
+var installationAnnotationAddCmd = &cobra.Command{
+	Use:   "add",
+	Short: "Adds Annotations to the Installation.",
```
Comment (gabrieljackson) [inline cmd/cloud/installation_annotation.go:35]:
```suggestion
	Short: "Adds annotations to the installation.",
```

File: cmd/cloud/cluster_annotation.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+var clusterAnnotationCmd = &cobra.Command{
+	Use:   "annotation",
+	Short: "Manipulate annotations of clusters managed by the provisioning server.",
+}
+
+var clusterAnnotationAddCmd = &cobra.Command{
+	Use:   "add",
+	Short: "Adds Annotations to the Cluster.",
```
Comment (gabrieljackson) [inline cmd/cloud/cluster_annotation.go:35]:
```suggestion
	Short: "Adds annotations to the cluster.",
```

File: cmd/cloud/installation_annotation.go [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+	Short: "Deletes Annotation from the Installation.",
+	RunE: func(command *cobra.Command, args []string) error {
+		command.SilenceUsage = true
+
+		serverAddress, _ := command.Flags().GetString("server")
+		client := model.NewClient(serverAddress)
+
+		installationID, _ := command.Flags().GetString("installation")
+		annotation, _ := command.Flags().GetString("annotation")
+
```
Comment (gabrieljackson) [inline cmd/cloud/installation_annotation.go:77]: Another request for the dry run flag if it makes sense.

File: internal/api/cluster.go [hunk 5]
```diff
@@ -576,3 +582,86 @@ func handleGetAllUtilityMetadata(c *Context, w http.ResponseWriter, r *http.Requ
...
+	clusterID := vars["cluster"]
+	c.Logger = c.Logger.WithField("cluster", clusterID).WithField("action", "add-cluster-annotations")
+
+	clusterDTO, status, unlockOnce := lockCluster(c, clusterID)
+	if status != 0 {
+		w.WriteHeader(status)
+		return
+	}
+	defer unlockOnce()
+
```
Comment (gabrieljackson) [inline internal/api/cluster.go:599]: We need to respect the API security lock like we do with other API endpoints.

This should be done for the other endpoints as well.

File: internal/api/cluster.go [hunk 6]
```diff
@@ -576,3 +582,86 @@ func handleGetAllUtilityMetadata(c *Context, w http.ResponseWriter, r *http.Requ
...
+// handleDeleteClusterAnnotation responds to DELETE /api/cluster/{cluster}/annotation/{annotation-name},
+// removes annotation from the Cluster.
+func handleDeleteClusterAnnotation(c *Context, w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	clusterID := vars["cluster"]
+	annotationName := vars["annotation-name"]
+	c.Logger = c.Logger.
+		WithField("cluster", clusterID).
+		WithField("action", "delete-cluster-annotation").
```
Comment (gabrieljackson) [inline internal/api/cluster.go:629]: Is there a reason we don't allow for deleting multiple annotations at the same time? I feel like it is a bit strange that we can create multiple annotations with one call, but have to delete them one-by-one. What do you think?

File: internal/store/annotation.go [hunk 7]
```diff
@@ -171,15 +195,80 @@ func (sqlStore *SQLStore) GetAnnotationsForClusters(filter *model.ClusterFilter)
 	return annotations, nil
 }
 
+// DeleteClusterAnnotation removes annotation from a given cluster.
```
Comment (gabrieljackson) [inline internal/store/annotation.go:198]:
```suggestion
// DeleteClusterAnnotation removes an annotation from a given cluster.
```

**Review by gabrieljackson — CHANGES_REQUESTED** [review]
This is a great improvement, but I have some questions.

File: internal/api/cluster.go (hunk 6 above)
Comment (Szymongib) [reply internal/api/cluster.go:629]: That results from API design. Following the REST convention, I am not a huge fan of passing body with the DELETE method therefor when deleting, the Annotation name is passed as a path variable.
GitHub API uses a similar approach with labels:
https://developer.github.com/v3/issues/labels/#remove-a-label-from-an-issue
https://developer.github.com/v3/issues/labels/#add-labels-to-an-issue

File: cmd/cloud/cluster_annotation.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	Short: "Deletes Annotation from the Cluster.",
+	RunE: func(command *cobra.Command, args []string) error {
+		command.SilenceUsage = true
+
+		serverAddress, _ := command.Flags().GetString("server")
+		client := model.NewClient(serverAddress)
+
+		clusterID, _ := command.Flags().GetString("cluster")
+		annotation, _ := command.Flags().GetString("annotation")
+
```
[the remark this answers is not in the document]
Comment (Szymongib) [reply cmd/cloud/cluster_annotation.go:77]: There is really nothing to dry run as it is sending the DELETE method without body. From what I have seen there are no `--dry-run` options for other commands which use DELETE underneath.

File: internal/api/cluster.go (hunk 5 above)
Comment (Szymongib) [reply internal/api/cluster.go:599]: Yes, that makes sense, I will add that.

File: internal/api/cluster.go (hunk 6 above)
Comment (gabrieljackson) [reply internal/api/cluster.go:635]: Okay this makes sense. Thanks for your reasoning.

In the future we may want to standardize a bit with our other API update endpoints, but we can discuss this as needed.

**Review by gabrieljackson — APPROVED** [review]
Awesome work!

**Pull request merged**